keps/sig-auth/3299-kms-v2-improvements/README.md

```diff
@@ -541,14 +578,81 @@ No.
 
 - 2022-05-09: Initial KEP draft submitted.
 - 2022-09-09: KMSv2 Alpha (v1.25) implemented https://github.com/kubernetes/kubernetes/pull/111126
+- 2023-02-28: re-use DEK while key ID is unchanged https://github.com/kubernetes/kubernetes/pull/116155
+  - This was a change in design from alpha to beta. The key hierarchy approach initially suggested in the KEP was dropped in favor of re-using DEKs based on metrics numbers from CI.
+- 2023-03-14: Generate proto API and update feature gate for beta https://github.com/kubernetes/kubernetes/pull/115123
 
 ## Alternatives
 
 **Performance and rotation:**
 
 We considered the follow approaches and each has its own drawbacks:
 1. `cacheSize` field in `EncryptionConfiguration`. It is used by the API server to initialize a LRU cache of the given size with the encrypted ciphertext used as index. Having a higher value for the `cacheSize` will prevent calls to the plugin for decryption operations. However, this does not solve the issue with the number of calls to KMS plugin when encryption traffic is bursty.
-1. Reduce the number of trips to KMS by caching DEKs by allowing one DEK to be used to encrypt multiple objects within the configured TTL period. One issue with this approach is it will be very hard to inform the API server to rotate the DEKs when a KEK has been rotated.
+2. Key hierarchy in the KMS plugin.
+   - No changes to the API server, keep 1:1 DEK mapping
+     - Assumption: A KMS plugin that was implemented using a local HSM would not need any changes because it would be able to handle the amount of encryption calls with ease since it would not need to perform network IO
+     - Assumption: local gRPC calls to the KMS plugin do not represent significant overhead
+   - KMS plugin generates its own local KEK in-memory
+   - External KMS is used to encrypt the local KEK
+   - Local KEK is used for encryption of DEKs sent by API server
+   - Local KEK is used for encryption based on policy (N events, X time, etc)
+  We tested this approach and the metrics in CI indicated the gRPC calls to the KMS plugin added significant overhead. The gRPC call latency was in the order of `0.1s` (refer to `apiserver_envelope_encryption_kms_operations_latency_seconds_bucket` [here](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/116148/pull-kubernetes-e2e-kind-kms/1630667339450028032/artifacts/metrics/kube-apiserver-metrics.txt))
```

Comment:
Lets copy these metrics into this PR and update this link since CI will eventually delete the file.

Reply:
If it is too large to include the whole file here lets use a GH gist to hold the complete set of metrics, and just include the encryption metrics here.